### PR TITLE
add as_noise_params

### LINKED ISSUE #187: create `as_noise_params()`
extract object construction code from `load_noise_params()`

### PR DESCRIPTION
**Pull request by chrisorwa**

### CODE REVIEW

File: tests/testthat/test-as_noise_params.R [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+      as_noise_params(),
+    class = "not noise_params"
+  )
+})
+
+test_that("`as_noise_params()` produces expected results", {
+  library(dplyr)
+  test_data <- "https://osf.io/download//hqy4v/" %>% # noise data
+    readr::read_rds() %>%
+    slice_head(n = 100) %>%
```
Comment (d-morrison) [inline tests/testthat/test-as_noise_params.R]: `slice_head()` not necessary when the original was only 16 rows.

**Review by d-morrison — CHANGES_REQUESTED** [review]
Looks great! Please address inline comment below, and bring Codecov patch coverage metric up to 100% (see bot comment).

File: R/as_noise_params.R [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+#'
+#' print(noise_data)
+#'
+as_noise_params <- function(data, antigen_isos = NULL) {
+  noise_data <-
+    data %>%
+    tibble::as_tibble()
+
+  # define noise columns
+  noise_cols <- c("antigen_iso", "Country", "y.low", "eps", "nu", "y.high")
```
Comment (d-morrison) [inline R/as_noise_params.R]: `Country` should not be required for `noise_params` objects, right @kaiemjoy?

File: R/as_noise_params.R [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  # get any missing column(s)
+  missing_cols <- setdiff(x = noise_cols, y = data_cols)
+
+  if (!all(is.element(noise_cols, noise_data %>% names()))) {
+    cli::cli_abort(
+      class = "not noise_params",
+      message = c(
+        "Can't convert {.arg data} to noise_params.",
+        "i" = "The column(s):{.strong {.var {missing_cols}}} are missing.",
+        "x" = "You have supplied {.cls {class(data)}}."
```
Comment (d-morrison) [inline R/as_noise_params.R]: why are we providing class information in this error message? it seems unrelated to the problem being checked for.

File: R/as_noise_params.R [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+  # assign curve class
+  class(noise_data) <-
+    c("noise_params", class(noise_data))
+
+  if (is.null(antigen_isos)) {
+    antigen_isos <- unique(noise_data$antigen_iso)
+  } else {
+    stopifnot(all(
+      is.element(antigen_isos, noise_data$antigen_iso)
```
Comment (d-morrison) [inline R/as_noise_params.R]: please expand this error message to inform the user which antigen isotypes are missing (probably using `cli_abort()` instead of `stopifnot()`)

File: tests/testthat/test-as_curve_params.R
Comment (d-morrison) [inline tests/testthat/test-as_curve_params.R:1]: It took me a while to figure out why edits to this file were included in a PR about `as_noise_params()`. I'm realizing now that it is to increase the PR test coverage. However, `as_curve_params.R` was only included in the PR test coverage metrics at all because of a deleted blank line; it would have been better to just undo that deletion and leave `as_curve_params.R` out of this PR altogether, unless there's a connection I'm missing. 

Going forward, please keep PRs simple and focused, so that they are easier to review.

File: R/as_noise_params.R [hunk 5]
```diff
@@ -32,7 +32,7 @@ as_noise_params <- function(data, antigen_isos = NULL) {
     cli::cli_abort(
       class = "not noise_params",
       message = c(
-        "Can't convert {.envvar data} to noise_params.",
+        "Can't convert {.arg data} to noise_params.",
         "i" = "The column(s):{.strong {.var {missing_cols}}} are missing.",
```
Comment (d-morrison) [inline R/as_noise_params.R]: Please use `cli`'s [pluralization](https://cli.r-lib.org/reference/pluralization.html) features (whenever we create or edit error messages):

- [x]  instead of `column(s)`, write `column{?s}`
- [x] instead of `are missing`, write `{?is/are} missing`

Also: 
- [x] markup `noise_params` as a `.cls`
- [x] put a space between the colon (`:`) and the opening bracket (`{`).

File: tests/testthat/test-as_noise_params.R [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  expect_snapshot_value(
+    x = test_data,
+    style = "serialize"
+  )
+})
+
+test_that("`as_noise_params()` produces error when wrong antigen_iso is provided", {
+  library(dplyr)
+
+  expect_error(object = "https://osf.io/download//hqy4v/" %>% # noise data
```
Comment (d-morrison) [inline tests/testthat/test-as_noise_params.R]: Please add an error class to the corresponding error and check for it here.

**Review by d-morrison — CHANGES_REQUESTED** [review]
Good work increasing the test coverage; please see inline comments.

File: R/as_noise_params.R (hunk 2 above)
Comment (chrisorwa) [reply R/as_noise_params.R]: @d-morrison we have `Country` in `noise_params` downloaded from OSF. 
image

File: R/as_noise_params.R (hunk 3 above)
Comment (chrisorwa) [reply R/as_noise_params.R]: The function `as_noise_params` can only convert a data frame to a `noise_params` class. If a user provides an `integer` or `character`, it should indicate so.

File: R/as_noise_params.R (hunk 4 above)
Comment (chrisorwa) [reply R/as_noise_params.R]: This is well-noted.

File: tests/testthat/test-as_noise_params.R (hunk 6 above)
Comment (chrisorwa) [reply tests/testthat/test-as_noise_params.R]: This is done.

File: R/as_noise_params.R (hunk 5 above)
Comment (chrisorwa) [reply R/as_noise_params.R]: This has been implemented.

**Comment by d-morrison** [comment]
@chrisorwa looks like this PR is back in draft mode; I'll take myself off review for now, please re-add me when ready.

File: R/as_curve_params.R [hunk 7]
```diff
@@ -1,8 +1,10 @@
-#' Load antibody decay curve parameter
+#' Load antibody decay curve parameters
 #'
 #' @param data a [data.frame()] or [tibble::tbl_df]
-#' @param antigen_isos [character()] vector of antigen isotypes to be used in analyses
-#' @returns a `curve_data` object (a [tibble::tbl_df] with extra attribute `antigen_isos`)
+#' @param antigen_isos [character()] vector of antigen
+#' isotypes to be used in analyses
+#' @returns a `curve_data` object (a [tibble::tbl_df]
+#' with extra attribute `antigen_isos`)
```
Comment (chrisorwa) [inline R/as_curve_params.R]: Limit sentences to 80 characters to meet linting requirement.

File: R/as_curve_params.R [hunk 8]
```diff
@@ -13,58 +15,46 @@
 #'
 #' print(curve_data)
 as_curve_params <- function(data, antigen_isos = NULL) {
-
-  if(!is.data.frame(data))
-  {
+  if (!is.data.frame(data)) {
```
Comment (chrisorwa) [inline R/as_curve_params.R]: Opening of curly braces not go on new line (lining requirement)

File: DESCRIPTION [hunk 9]
```diff
@@ -1,7 +1,7 @@
 Package: serocalculator
 Type: Package
 Title: Estimating Infection Rates from Serological Data
-Version: 1.2.0.9001
+Version: 1.3.0
```
Comment (chrisorwa) [inline DESCRIPTION]: Increment version (the version number conflicts with PR #234 ) - this will depend upon which one gets merged first and then the increment version will be updated.

File: NAMESPACE [hunk 10]
```diff
@@ -28,6 +28,7 @@ S3method(summary,pop_data)
 S3method(summary,seroincidence)
 S3method(summary,seroincidence.by)
 export(as_curve_params)
+export(as_noise_params)
```
Comment (chrisorwa) [inline NAMESPACE:18]: Auto-generated by the build process.

File: NEWS.md [hunk 11]
```diff
@@ -1,4 +1,6 @@
-# serocalculator (development version)
+# serocalculator 1.3.0
+
+* add `as_noise_params`
```
Comment (chrisorwa) [inline NEWS.md]: Version increment with the added file/functionality.

File: R/as_curve_params.R (hunk 8 above)
Comment (chrisorwa) [inline R/as_curve_params.R]: Linting required opening curly brackets not to be on a start a new line.

File: R/as_curve_params.R [hunk 12]
```diff
@@ -13,58 +15,46 @@
...
     cli::cli_abort(
       class = "not data.frame",
       message = c(
         "Can't convert {.arg data} to {.cls curve_params}.",
-        "x" = "{.arg data} must be a {.cls data.frame}
-        (or a subclass of {.cls data.frame}).",
+        "x" = "{.arg data} must be a {.cls data.frame}" %>%
+          "(or a subclass of {.cls data.frame}).",
```
Comment (chrisorwa) [inline R/as_curve_params.R]: Break the string into two strings to fit 80 characters per line (linting).

File: R/as_noise_params.R [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    cli::cli_abort(
+      class = "not data.frame",
+      message = c(
+        "Can't convert {.arg data} to {.cls noise_params}.",
+        paste("x" = "{.arg data} must be a {.cls data.frame}",
+          "(or a subclass of {.cls data.frame})."
+        ),
+        "i" = "You have supplied a {.cls {class(data)}}."
+      )
+    )
```
Comment (chrisorwa) [inline R/as_noise_params.R]: Convert data frame to `noise_params` class.

File: R/est.incidence.by.R [hunk 14]
```diff
@@ -1,14 +1,22 @@
...
+#' @param strata a [character] vector of stratum-defining variables.
+#' Values must be variable names in `pop_data`.
+#' @param curve_strata_varnames A subset of `strata`.
+#' Values must be variable names in `curve_params`. Default = "".
+#' @param noise_strata_varnames A subset of `strata`.
+#' Values must be variable names in `noise_params`. Default = "".
+#' @param num_cores Number of processor cores to use for
+#' calculations when computing by strata. If set to
+#' more than 1 and package \pkg{parallel} is available,
+#' then the computations are executed in parallel. Default = 1L.
```
Comment (chrisorwa) [inline R/est.incidence.by.R]: Linting to 80 characters

File: R/est.incidence.by.R [hunk 15]
```diff
@@ -17,7 +25,8 @@
 #' and then the data will be passed to [est.incidence()].
 #' If for some reason you want to use [est.incidence.by()]
 #' with no strata instead of calling [est.incidence()],
-#' you may use `NA`, `NULL`, or `""` as the `strata` argument to avoid that warning.
+#' you may use `NA`, `NULL`, or `""` as the `strata`
+#' argument to avoid that warning.
```
Comment (chrisorwa) [inline R/est.incidence.by.R]: Lint to 80 characters.

File: R/est.incidence.by.R [hunk 16]
```diff
@@ -26,7 +35,9 @@
 #'
 #' @return
 #' * if `strata` has meaningful inputs:
-#' An object of class `"seroincidence.by"`; i.e., a list of `"seroincidence"` objects from [est.incidence()], one for each stratum, with some meta-data attributes.
+#' An object of class `"seroincidence.by"`; i.e., a list of `
+#' "seroincidence"` objects from [est.incidence()], one for each stratum,
+#' with some meta-data attributes.
```
Comment (chrisorwa) [inline R/est.incidence.by.R]: Linting to 80 characters.

File: R/est.incidence.by.R [hunk 17]
```diff
@@ -39,7 +50,7 @@
 #'
 #' curve <- load_curve_params("https://osf.io/download/rtw5k/") %>%
 #'   filter(antigen_iso %in% c("HlyE_IgA", "HlyE_IgG")) %>%
-#'   slice(1:100, .by = antigen_iso) # Reduce dataset for the purposes of this example
+#'   slice(1:100, .by = antigen_iso)
```
Comment (chrisorwa) [inline R/est.incidence.by.R]: Remove comment to meet 80 character linting limit.

File: R/est.incidence.by.R [hunk 18]
```diff
@@ -49,55 +60,58 @@
...
+  antigen_isos = pop_data %>%
+    pull("antigen_iso") %>%
+    unique(),
+  lambda_start = 0.1,
+  build_graph = FALSE,
+  num_cores = 1L,
+  verbose = FALSE,
+  print_graph = FALSE,
+  ...
+) {
```
Comment (chrisorwa) [inline R/est.incidence.by.R]: Linting to meet spacing requirement

File: R/est.incidence.by.R [hunk 19]
```diff
@@ -49,55 +60,58 @@
...
     warning(
       "The `strata` argument to `est.incidence.by()` is missing.",
       "\n\n  If you do not want to stratify your data, ",
-      "consider using the `est.incidence()` function to simplify your code and avoid this warning.",
-      "\n\n Since the `strata` argument is empty, `est.incidence.by()` will return a `seroincidence` object, instead of a `seroincidence.by` object.\n"
+      "consider using the `est.incidence()` function to",
+      "simplify your code and avoid this warning.",
+      "\n\n Since the `strata` argument is empty, `est.incidence.by()`",
+      "will return a `seroincidence` object, instead of a",
+      "`seroincidence.by` object.\n"
```
Comment (chrisorwa) [inline R/est.incidence.by.R]: Reduce sentences to 80 characters per line to meet linting requirement.

File: R/est.incidence.by.R [hunk 20]
```diff
@@ -49,55 +60,58 @@
...
   }
 
   strata_is_empty <-
     missing(strata) ||
-      is.null(strata) ||
-      setequal(strata, NA) ||
-      setequal(strata, "")
+    is.null(strata) ||
+    setequal(strata, NA) ||
+    setequal(strata, "")
```
Comment (chrisorwa) [inline R/est.incidence.by.R]: Reduce spacing to meet linting requirement.

File: tests/testthat/test-as_noise_params.R [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+          when non-noise data is provided", {
+            library(magrittr)
+            expect_error(
+              object = noise_data <-
+                "https://osf.io/download//n6cp3/" %>% # pop data
+                readr::read_rds() %>%
+                as_noise_params(),
+              class = "not noise_params"
+            )
+          })
```
Comment (chrisorwa) [inline tests/testthat/test-as_noise_params.R:11]: Test that `as_noise_params` only process `noise_params` data. Pass `pop_data` to check it `as_noise_params` will raise an error.

File: tests/testthat/test-as_noise_params.R [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+    readr::read_rds() %>%
+    as_noise_params()
+
+  expect_snapshot(test_data)
+
+  expect_snapshot_value(
+    x = test_data,
+    style = "serialize"
+  )
+})
```
Comment (chrisorwa) [inline tests/testthat/test-as_noise_params.R:25]: Pass `noise_params` data and test if `as_noise_params` will process it without throwing an error.

File: tests/testthat/test-as_noise_params.R [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+          wrong antigen_iso is provided", {
+            library(dplyr)
+
+            expect_error(
+              object = "https://osf.io/download//hqy4v/" %>% # noise data
+                readr::read_rds() %>%
+                as_noise_params(antigen_iso = "HlyE_IgB"),
+              class = "missing-antigen"
+            )
+          })
```
Comment (chrisorwa) [inline tests/testthat/test-as_noise_params.R:37]: Test if  `as_noise_params` raise an error when provided `antigen_iso`  does not exist in the data.

File: R/load_noise_params.R [hunk 24]
```diff
@@ -17,18 +22,7 @@ load_noise_params <- function(file_path, antigen_isos = NULL) {
...
-  class(noise) <-
-    c("noise", class(noise))
-
-  if (is.null(antigen_isos)) {
-    antigen_isos <- unique(noise$antigen_iso)
-  } else {
-    stopifnot(all(is.element(antigen_isos, noise$antigen_iso)))
-  }
-
-  attr(noise, "antigen_isos") <- antigen_isos
```
Comment (chrisorwa) [inline R/load_noise_params.R:31]: Functionality moved to `as_noise_params`

File: R/load_noise_params.R [hunk 25]
```diff
@@ -1,9 +1,14 @@
 #' Load noise parameters
 #'
-#' @param file_path path to an RDS file containing biologic and measurement noise of antibody decay curve parameters `y.low`, `eps`, `nu`, and `y.high`, stored as a [data.frame()] or [tibble::tbl_df]
-#' @param antigen_isos [character()] vector of antigen isotypes to be used in analyses
+#' @param file_path path to an RDS file containing biologic
+#' and measurement noise of antibody decay curve parameters
+#' `y.low`, `eps`, `nu`, and `y.high`,
+#' stored as a [data.frame()] or [tibble::tbl_df]
+#' @param antigen_isos [character()] vector
+#' of antigen isotypes to be used in analyses
```
Comment (chrisorwa) [inline R/load_noise_params.R:8]: Reduce sentences to 80 characters (linting requirement)

File: R/load_noise_params.R [hunk 26]
```diff
@@ -1,9 +1,14 @@
...
 #'
-#' @returns a `noise` object (a [tibble::tbl_df] with extra attribute `antigen_isos`)
+#' @returns a `noise` object (a [tibble::tbl_df]
+#' with extra attribute `antigen_isos`)
```
Comment (chrisorwa) [inline R/load_noise_params.R:11]: Reduce sentence to 80 characters (linting requirement).

File: R/est.incidence.by.R [hunk 27]
```diff
@@ -142,90 +158,80 @@ est.incidence.by <- function(
...
+          list(
+            lambda_start = lambda_start,
+            antigen_isos = antigen_isos,
+            build_graph = build_graph,
+            print_graph = print_graph,
+            verbose = verbose,
+            ...
+          )
+        )
+      )
```
Comment (chrisorwa) [inline R/est.incidence.by.R]: Changing how the code operates to avoid `->` assignment which bring up a warning on linting. Assignment should right to left.

File: R/est.incidence.by.R [hunk 28]
```diff
@@ -142,90 +158,80 @@ est.incidence.by <- function(
...
+              verbose = FALSE,
+              ...
             )
           )
-        }
-      )
-    } %>% system.time() -> time
+        )
+      }
+    ))
```
Comment (chrisorwa) [inline R/est.incidence.by.R]: Changing `->` assignment to `<-`

File: R/est.incidence.by.R [hunk 29]
```diff
@@ -142,90 +158,80 @@ est.incidence.by <- function(
...
+          what = est.incidence,
+          args = c(
+            x,
+            list(
+              lambda_start = lambda_start,
+              antigen_isos = antigen_isos,
+              build_graph = build_graph,
+              print_graph = FALSE,
+              verbose = FALSE,
+              ...
```
Comment (chrisorwa) [inline R/est.incidence.by.R]: Redo the code to avoid `<-` object value assignment.

File: R/est.incidence.by.R [hunk 30]
```diff
@@ -142,90 +158,80 @@ est.incidence.by <- function(
     num_cores <- num_cores %>% check_parallel_cores()
 
     if (verbose) {
-      message("Setting up parallel processing with `num_cores` = ", num_cores, ".")
+      message("Setting up parallel processing with",
+              "`num_cores` = ", num_cores, ".")
```
Comment (chrisorwa) [inline R/est.incidence.by.R]: Reduce code to 80 characters long

File: R/est.incidence.by.R [hunk 31]
```diff
@@ -130,8 +144,10 @@ est.incidence.by <- function(
 
   if (num_cores > 1L && !requireNamespace("parallel", quietly = TRUE)) {
     warning(
-      "The `parallel` package is not installed, so `num_cores > 1` has no effect.",
-      "To install `parallel`, run `install.packages('parallel')` in the console."
+      "The `parallel` package is not installed,",
+      "so `num_cores > 1` has no effect.",
+      "To install `parallel`, run `install.packages('parallel')`",
+      "in the console."
```
Comment (chrisorwa) [inline R/est.incidence.by.R]: Reduce sentences to 80 characters.

File: R/as_curve_params.R [hunk 32]
```diff
@@ -13,58 +15,47 @@
...
-  # check if object is curve (with columns)
-  if (!all(is.element(curve_cols, curve_data %>% names()))) {
-    # get columns from provided data
-    data_cols <- data %>% names()
-
-    # get any missing column(s)
-    missing_cols <- setdiff(x = curve_cols, y = data_cols)
-
+  # Check if object is a curve (with columns)
+  if (!all(is.element(curve_cols, names(curve_data)))) {
```
Comment (chrisorwa) [inline R/as_curve_params.R]: Moving getting columns into `is.element` function. Reduce unnecessary objects.

File: tests/testthat/test-as_noise_params.R (hunk 1 above)
Comment (chrisorwa) [reply tests/testthat/test-as_noise_params.R]: This has been removed.

File: tests/testthat/test-as_curve_params.R
Comment (chrisorwa) [reply tests/testthat/test-as_curve_params.R:1]: We can keep the additional requests to the PR to avoid scope creep.

**Review by d-morrison — CHANGES_REQUESTED** [review]
Not sure I understand how this PR ended up containing the other ones in it; please reduce it to just the changes relevant to #187?

File: NEWS.md [hunk 33]
```diff
@@ -1,4 +1,6 @@
-# serocalculator (development version)
+# serocalculator 1.3.0
```
Comment (d-morrison) [inline NEWS.md]: please revert the first line of NEWS.md; this isn't a release yet, so the header should stay as `# serocalculator (development version)` for now.

File: NEWS.md [hunk 34]
```diff
@@ -1,4 +1,6 @@
...
+
+* Add `as_noise_params`
```
Comment (d-morrison) [inline NEWS.md]: please add the PR number in parentheses; i.e., `(#228)`

File: R/as_noise_params.R (hunk 13 above)
Comment (d-morrison) [inline R/as_noise_params.R]: please add a unit test covering these lines.

File: R/as_noise_params.R (hunk 3 above)
Comment (d-morrison) [reply R/as_noise_params.R]: we moved the class-checking to lines 21-28, so we don't need line 49 here; please remove.

**Review by d-morrison — CHANGES_REQUESTED** [review]
Nearly there! Please see comments.

File: tests/testthat/_snaps/est.incidence.by.md
Comment (d-morrison) [inline tests/testthat/_snaps/est.incidence.by.md:1]: See below; seems like this should be in a separate PR.

File: tests/testthat/test-est.incidence.by.R
Comment (d-morrison) [inline tests/testthat/test-est.incidence.by.R:1]: why is this test file included in this PR? It seems like it should be a separate PR.

File: NEWS.md [hunk 35]
```diff
@@ -1,5 +1,10 @@
 # serocalculator (development version)
 
+# serocalculator 1.2.9.9007
+* Add `as_noise_params` (#228) 
+
+* Updated enteric fever example article with upgraded code and visualizations
```
Comment (d-morrison) [inline NEWS.md]: I don't see any changes to the enteric fever example article in this PR, and I don't think that we need them to be in this PR; did this line get accidentally added from a different PR?

**Review by d-morrison — CHANGES_REQUESTED** [review]
Thanks for addressing my previous comments! In the process, it looks like some new unrelated changes got accidentally introduced; please move them to a separate PR.

File: R/as_noise_params.R [hunk 36]
```diff
@@ -0,0 +1,78 @@
...
+
+  # Get any missing columns
+  missing_cols <- setdiff(x = noise_cols, y = names(data))
+
+  if (length(missing_cols) > 0) {
+    cli::cli_abort(
+      class = "not noise_params",
+      message = c(
+        "Can't convert {.arg data} to {.cls noise_params}.",
+        "i" = paste("The column{?s} : {.strong {.var {missing_cols}}}",
```
Comment (d-morrison) [inline R/as_noise_params.R]: there's an extra space **before** the colon here (keep the one **after** the colon)

Comment (chrisorwa) [reply R/as_noise_params.R]: This is well-noted. Do note that the `styleR` library does not put spaces before or after the colon.

```
  if (length(missing_cols) > 0) {
    cli::cli_abort(
      class = "not noise_params",
      message = c(
        "Can't convert {.arg data} to {.cls noise_params}.",
        "i" = paste(
          "The column{?s}:{.strong {.var {missing_cols}}}",
          "{?is/are} missing."
        )
      )
    )
  }
   ```

File: NEWS.md (hunk 35 above)
Comment (chrisorwa) [reply NEWS.md]: The line got added to this PR when the branch was updated.

File: NEWS.md [hunk 37]
```diff
@@ -1,5 +1,9 @@
 # serocalculator (development version)
 
+# serocalculator 1.2.9.9007
```
Comment (d-morrison) [inline NEWS.md]: no need to add this header

File: NEWS.md [hunk 38]
```diff
@@ -1,5 +1,9 @@
...
+
+* Add `as_noise_params` (#228)
```
Comment (d-morrison) [inline NEWS.md:5]: Please move this line after line 8 (under `## New features`)

File: R/as_noise_params.R (hunk 36 above)
Comment (d-morrison) [reply R/as_noise_params.R]: Does `styler` even attempt to edit strings? I'd guess not.
Even if it does - it's fine to use `styler` or other tools to help clean up code, but please still do a final, detailed visual edit before requesting review.

**Review by d-morrison — CHANGES_REQUESTED** [review]
Looks great - please make two last changes to `NEWS.md` (see comments), and we can merge this.

File: NEWS.md (hunk 38 above)
Comment (chrisorwa) [reply NEWS.md:5]: This has been implemented. 
image

Comment (chrisorwa) [reply NEWS.md:5]: This has been done.

File: R/as_noise_params.R [hunk 39]
```diff
@@ -0,0 +1,82 @@
...
+      class = "not noise_params",
+      message = c(
+        "Can't convert {.arg data} to {.cls noise_params}.",
+        "i" = paste(
+          "The column{?s}: {.strong {.var {missing_cols}}}",
+          "{?is/are} missing."
+        )
+      )
+    )
+  }
```
Comment (chrisorwa) [inline R/as_noise_params.R:53]: The space before the colon is removed, and only the space after is retained.

File: NEWS.md (hunk 37 above)
Comment (chrisorwa) [reply NEWS.md]: This is well-noted.

File: R/as_noise_params.R (hunk 36 above)
Comment (chrisorwa) [reply R/as_noise_params.R]: `styleR` does edit strings. The space before the colon has been removed.

**Review by d-morrison — APPROVED** [review]
Looks great - thanks!
